File: StatisticalAssumptions/assumptionsRegressions.r
### Linearity 
p1 <- ggplot(lin_reg, aes(.fitted, .resid)) + geom_point()
p1 <- p1 + stat_smooth(method="loess") + geom_hline(yintercept=0, col="red", linetype="dashed")
p1 <- p1 + xlab("Predicted") + ylab("Residuals")
p1 <- p1 + ggtitle("Residuals vs. Predicted Values") + theme_bw()

df_plt <- data.frame("fitted" = fitted(lin_reg), "observed" = X$medv)
p2 <- ggplot(df_plt, aes(x=fitted, y=observed)) + geom_point()
p2 <- p2 + stat_smooth(method="loess") + geom_abline(intercept = 1, col="red", linetype="dashed")
p2 <- p2 + xlab("Predicted") + ylab("Observed")
p2 <- p2 + ggtitle("Observed vs. Predicted Values") + theme_bw()

grid.arrange(p2, p1, ncol=2)




### Expectation (mean) of residuals is zero
mean(lin_reg$resid)



### No (perfect) multicollinearity
library(car)
vif(lin_reg)



#### Homoscedasticity (equal variance) of residuals
library(lmtest)

par(mfrow=c(2,2))  # set 2 rows and 2 column plot layout
plot(lin_reg)
print(bptest(lin_reg, data = X, studentize = TRUE)) # Breusch-Pagan test
print(gqtest(lin_reg)) # Goldfeld-Quandt



### No autocorrelation of residuals
library(ggplot2)
library(lmtest)

acf(lin_reg$residuals)  
dwtest(lin_reg)



### The features and residuals are uncorrelated
for (i in 1:(dim(X)[2])){
    cor_test <- cor.test(X[, i], lin_reg$residuals)  # 
    print(paste('Variable:', colnames(X)[i], 
                '--- correlation:', as.character(cor_test$estimate), 
                ', p-value:', as.character(cor_test$p.value), sep = " ", collapse = NULL))
    }


### There must be some variability in features
library(caret)
apply(X, 2, var)
nearZeroVar(X, saveMetrics= TRUE)

File: StatisticalAssumptions/assumptionsRegressions.py


## Linearity of the model
## Linearity of the model
## Linearity of the model
## Linearity of the model


# %matplotlib inline
# %config InlineBackend.figure_format ='retina'
import seaborn as sns 
import matplotlib.pyplot as plt
import statsmodels.stats.api as sms
sns.set_style('darkgrid')
sns.mpl.rcParams['figure.figsize'] = (15.0, 9.0)

def linearity_test(model, y):
    '''
    Function for visually inspecting the assumption of linearity in a linear regression model.
    It plots observed vs. predicted values and residuals vs. predicted values.
    
    Args:
    * model - fitted OLS model from statsmodels
    * y - observed values
    '''
    fitted_vals = model.predict()
    resids = model.resid

    fig, ax = plt.subplots(1,2)
    
    sns.regplot(x=fitted_vals, y=y, lowess=True, ax=ax[0], line_kws={'color': 'red'})
    ax[0].set_title('Observed vs. Predicted Values', fontsize=16)
    ax[0].set(xlabel='Predicted', ylabel='Observed')

    sns.regplot(x=fitted_vals, y=resids, lowess=True, ax=ax[1], line_kws={'color': 'red'})
    ax[1].set_title('Residuals vs. Predicted Values', fontsize=16)
    ax[1].set(xlabel='Predicted', ylabel='Residuals')
    
linearity_test(lin_reg, y)   





### Expectation (mean) of residuals is zero
### Expectation (mean) of residuals is zero
### Expectation (mean) of residuals is zero
### Expectation (mean) of residuals is zero

lin_reg.resid.mean()








### No (perfect) multicollinearity
### No (perfect) multicollinearity
### No (perfect) multicollinearity
### No (perfect) multicollinearity


from statsmodels.stats.outliers_influence import variance_inflation_factor

vif = [variance_inflation_factor(X_constant.values, i) for i in range(X_constant.shape[1])]
pd.DataFrame({'vif': vif[1:]}, index=X.columns).T






#### Homoscedasticity (equal variance) of residuals
#### Homoscedasticity (equal variance) of residuals
#### Homoscedasticity (equal variance) of residuals
#### Homoscedasticity (equal variance) of residuals
#### Homoscedasticity (equal variance) of residuals

# %matplotlib inline
# %config InlineBackend.figure_format ='retina'
import seaborn as sns 
import matplotlib.pyplot as plt
import statsmodels.stats.api as sms
sns.set_style('darkgrid')
sns.mpl.rcParams['figure.figsize'] = (15.0, 9.0)

def homoscedasticity_test(model):
    '''
    Function for testing the homoscedasticity of residuals in a linear regression model.
    It plots residuals and standardized residuals vs. fitted values and runs Breusch-Pagan and Goldfeld-Quandt tests.
    
    Args:
    * model - fitted OLS model from statsmodels
    '''
    fitted_vals = model.predict()
    resids = model.resid
    resids_standardized = model.get_influence().resid_studentized_internal

    fig, ax = plt.subplots(1,2)

    sns.regplot(x=fitted_vals, y=resids, lowess=True, ax=ax[0], line_kws={'color': 'red'})
    ax[0].set_title('Residuals vs Fitted', fontsize=16)
    ax[0].set(xlabel='Fitted Values', ylabel='Residuals')

    sns.regplot(x=fitted_vals, y=np.sqrt(np.abs(resids_standardized)), lowess=True, ax=ax[1], line_kws={'color': 'red'})
    ax[1].set_title('Scale-Location', fontsize=16)
    ax[1].set(xlabel='Fitted Values', ylabel='sqrt(abs(Residuals))')

    bp_test = pd.DataFrame(sms.het_breuschpagan(resids, model.model.exog), 
                           columns=['value'],
                           index=['Lagrange multiplier statistic', 'p-value', 'f-value', 'f p-value'])

    gq_test = pd.DataFrame(sms.het_goldfeldquandt(resids, model.model.exog)[:-1],
                           columns=['value'],
                           index=['F statistic', 'p-value'])

    print('\n Breusch-Pagan test ----')
    print(bp_test)
    print('\n Goldfeld-Quandt test ----')
    print(gq_test)
    print('\n Residuals plots ----')

homoscedasticity_test(lin_reg)






### The features and residuals are uncorrelated
### The features and residuals are uncorrelated
### The features and residuals are uncorrelated
### The features and residuals are uncorrelated

from scipy.stats.stats import pearsonr

for column in X.columns:
    corr_test = pearsonr(X[column], lin_reg.resid)
    print(f'Variable: {column} --- correlation: {corr_test[0]:.4f}, p-value: {corr_test[1]:.4f}')




### The number of observations must be greater than the number of features


### There must be some variability in features
X.apply(np.var, axis=0)